### PR TITLE
Enable flow analysis based rules for non-method body executable code

### LINKED ISSUE #1650: Support for dataflow analysis for non-method body executable code
Our current DFA/CFG implementation only supports flow analysis for block operations. This issue tracks adding support in non-method body executable code such as initializers.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #1650

While fixing this I ran into other bugs related to flow analysis, which have been fixed as well.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidDeadConditionalCode.cs [hunk 1]
```diff
@@ -72,13 +70,20 @@ public override void Initialize(AnalysisContext context)
 
                         if (operationRoot.HasAnyOperationDescendant(ShouldAnalyze))
                         {
-                            var cfg = operationRoot.GetEnclosingControlFlowGraph();
+                            // Skip duplicate analysis from operation blocks for constructor initializer and body.
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidDeadConditionalCode.cs]: This is unfortunate side effect of the fact that we have different OperationBlocks for constructor initializer and constructor method body, but both of them are part of the same control flow graph. However, most analyzer authors should not require such duplicate checking, as unlike this analyzer which performs flow analysis on every operation block, most analyzers will do so only on specific operations.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_StringContentAnalysis.cs [hunk 2]
```diff
@@ -869,5 +869,79 @@ End Sub
...
+
+class Base
+{
+    protected Base(bool b) { }
+}
+
+class Test : Base
+{
+    public Test(string s1, string s2, string s3)
+        : base(s1 == s2 && s2 == s3 ? (s1 == s3) : false)
```
Comment (333fred) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_StringContentAnalysis.cs]: >s1 == s3 [](start = 39, length = 8)

I assume that this is where the warning actually is? Can you add a `/**/` comment of some kind so it's more easy to visually differentiate? I spent about 20 minutes attempting to figure out how the lower `s1 == s3` was provably true.

Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_StringContentAnalysis.cs]: Yes, and sorry :-) The lower line in the body is just to verify that any data derived from predicates/conditions in the initializer is not carried forward into the method body.

**Comment by mavasani** [comment]
Ping for reviews...

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_StringContentAnalysis.cs (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_StringContentAnalysis.cs]: I will update the tests with next change in this area.

**Pull request merged**